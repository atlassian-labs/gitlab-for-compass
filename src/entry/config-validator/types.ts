export interface ForgeTriggerContext {
  installContext: string;
}
