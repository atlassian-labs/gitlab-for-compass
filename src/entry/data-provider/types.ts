type DataProviderPayload = {
  url: string;
  ctx: {
    cloudId: string;
    extensionId: string;
  };
};

type CallbackPayload = {
  success: boolean;
  url: string;
  errorMessage?: string;
};

export { DataProviderPayload, CallbackPayload };
