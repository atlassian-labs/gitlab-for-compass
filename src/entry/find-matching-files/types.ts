export type FindMatchingFilesPayload = {
  repoUrl: string;
  fileName: {
    // These are OR'd together
    equalsOneOf: string[];
    containsOneOf: string[];
  };
  context: {
    cloudId: string;
    extensionId: string;
  };
};

export type File = {
  path: string;
  metadata: {
    size?: number;
  };
};

export type FindMatchingFilesResponse = {
  success: boolean;
  statusCode: number;
  files: File[];
  errorMessage?: string;
};
