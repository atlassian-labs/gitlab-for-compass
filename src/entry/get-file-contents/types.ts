type GetFileContentsPayload = {
  projectId: number;
  groupId: number;
  filePath: string;
  ref?: string;
};

type GetFileContentsResponse = {
  success: boolean;
  statusCode: number;
  errorMessage?: string;
  file?: {
    contents: string;
  };
};

export { GetFileContentsPayload, GetFileContentsResponse };
