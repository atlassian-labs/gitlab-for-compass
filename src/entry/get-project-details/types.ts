type GetProjectDetailsPayload = {
  projectUrl: string;
};

type GetProjectDetailsResponse = {
  success: boolean;
  statusCode: number;
  errorMessage?: string;
  project?: {
    projectId: number;
    groupId: number;
    defaultBranch: string;
    shaOnDefaultBranch: string;
  };
};

export { GetProjectDetailsPayload, GetProjectDetailsResponse };
