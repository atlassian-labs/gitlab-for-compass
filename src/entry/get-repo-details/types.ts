type GetRepoDetailsPayload = {
  projectUrl: string;
};

type GetRepoDetailsResponse = {
  success: boolean;
  statusCode: number;
  errorMessage?: string;
  project?: {
    projectId: number;
    groupId: number;
    defaultBranch: string;
    shaOnDefaultBranch: string;
  };
};

export { GetRepoDetailsPayload, GetRepoDetailsResponse };
