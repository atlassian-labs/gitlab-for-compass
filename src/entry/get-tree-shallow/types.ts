// https://docs.gitlab.com/api/repositories/#list-repository-tree
import { ProjectFile } from '../../types';

type GetTreeShallowPayload = {
  projectId: number;
  groupId: number;
  ref?: string;
  path?: string;
  pageToken?: string;
  pageSize?: number;
  recursive?: boolean;
};

type GetTreeShallowResponse = {
  success: boolean;
  statusCode: number;
  files?: ProjectFile[];
  errorMessage?: string;
};

export { GetTreeShallowPayload, GetTreeShallowResponse };
