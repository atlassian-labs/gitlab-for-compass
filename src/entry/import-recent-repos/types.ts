export type ImportRecentReposPayload = {
  cloudId: string;
  numRepos: number;
};

export type ImportRecentReposReturn = {
  success: boolean;
  errors?: string;
  response: {
    numReposImported: number;
  };
};

export type ImportResultsSummary = {
  successfulImports: number;
  failedImports: number;
};
