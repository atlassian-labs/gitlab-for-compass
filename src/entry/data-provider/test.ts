/* eslint-disable import/first, import/order */
import { mockAgg } from '../../__tests__/helpers/mock-agg';

mockAgg();

import {
  DataProviderBuildEvent,
  DataProviderDeploymentEvent,
  CompassBuildEventState,
  CompassDeploymentEventEnvironmentCategory,
  CompassDeploymentEventState,
} from '@atlassian/forge-graphql';
import { dataProvider } from './index';
import * as getBackfillEvents from '../../services/get-backfill-data';
import * as getProjectDataFromUrl from '../../services/data-provider-link-parser';
import * as getTrackingBranchName from '../../services/get-tracking-branch';
import { GitlabAPIProject } from 'src/types';

const getEventsSpy = jest.spyOn(getBackfillEvents, 'getBackfillData');
const projectDataSpy = jest.spyOn(getProjectDataFromUrl, 'getProjectDataFromUrl');
const trackingBranchSpy = jest.spyOn(getTrackingBranchName, 'getTrackingBranchName');

const MOCK_BUILD_EVENT: DataProviderBuildEvent = {
  pipeline: {
    pipelineId: 'pipeline-id',
  },
  startedAt: 'started-at',
  completedAt: 'completed-at',
  state: CompassBuildEventState.Successful,
  description: 'mock description',
  displayName: 'mock display name',
  updateSequenceNumber: 1,
  lastUpdated: 'updated',
  url: 'url',
};

export const MOCK_DEPLOY_EVENT: DataProviderDeploymentEvent = {
  displayName: 'name',
  lastUpdated: 'mock',
  updateSequenceNumber: '1',
  url: 'url',
  environment: {
    category: CompassDeploymentEventEnvironmentCategory.Production,
    displayName: 'prod',
    environmentId: 'id',
  },
  pipeline: { pipelineId: '1', displayName: 'pipeline', url: 'url' },
  sequenceNumber: 1,
  state: CompassDeploymentEventState.Successful,
};

const MOCK_EVENTS_RESPONSE: {
  builds: DataProviderBuildEvent[];
  deployments: DataProviderDeploymentEvent[];
  metrics: {
    mrCycleTime: number;
    buildDuration: number;
    openMergeRequestsCount: number;
  };
} = {
  builds: [MOCK_BUILD_EVENT],
  deployments: [MOCK_DEPLOY_EVENT],
  metrics: {
    mrCycleTime: 1,
    buildDuration: 2,
    openMergeRequestsCount: 3,
  },
};
const MOCK_PROJECT_URL = 'https://gitlab.com/test/repo-name?testParam=test';
const MOCK_PROJECT: GitlabAPIProject = {
  id: 1,
  description: 'description',
  name: 'name',
  topics: ['topic'],
  default_branch: 'main',
  web_url: 'web_url',
  namespace: {
    id: 2,
    full_path: 'full_path',
    name: 'name',
    path: 'path',
  },
  created_at: 'abc',
};

describe('dataProvider module', () => {
  it('successfully returns events and metrics in the expected format', async () => {
    getEventsSpy.mockResolvedValue(MOCK_EVENTS_RESPONSE);
    projectDataSpy.mockResolvedValue({
      project: MOCK_PROJECT,
      groupToken: 'mock-group-token',
    });

    trackingBranchSpy.mockResolvedValue('branch');

    const result = await dataProvider({
      url: MOCK_PROJECT_URL,
      ctx: {
        cloudId: 'ari:cloud:compass:122345:component/12345/12345',
        extensionId: 'mock-extension-id',
      },
    });

    expect(result.externalSourceId).toEqual(MOCK_PROJECT.id.toString());
    expect(result.metrics).toMatchSnapshot();
    expect(result.events).toMatchSnapshot();
  });
});
