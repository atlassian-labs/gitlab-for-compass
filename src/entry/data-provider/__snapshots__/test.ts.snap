// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataProvider module successfully returns events and metrics in the expected format 1`] = `
Object {
  "builtIn": Object {
    "ari:cloud:compass::metric-definition/builtin/build-success-rate": Object {
      "derived": true,
      "initialValue": null,
    },
    "ari:cloud:compass::metric-definition/builtin/build-time-avg-last-10": Object {
      "derived": true,
      "initialValue": null,
    },
    "ari:cloud:compass::metric-definition/builtin/open-pull-requests": Object {
      "initialValue": 3,
    },
    "ari:cloud:compass::metric-definition/builtin/pull-request-cycle-time-avg-last-10": Object {
      "initialValue": 1,
    },
    "ari:cloud:compass::metric-definition/builtin/weekly-deployment-frequency-28d": Object {
      "derived": true,
      "initialValue": null,
    },
  },
  "custom": undefined,
}
`;

exports[`dataProvider module successfully returns events and metrics in the expected format 2`] = `
Object {
  "builds": Object {
    "initialValues": Array [
      Object {
        "completedAt": "completed-at",
        "description": "mock description",
        "displayName": "mock display name",
        "lastUpdated": "updated",
        "pipeline": Object {
          "pipelineId": "pipeline-id",
        },
        "startedAt": "started-at",
        "state": "SUCCESSFUL",
        "updateSequenceNumber": 1,
        "url": "url",
      },
    ],
  },
  "deployments": Object {
    "initialValues": Array [
      Object {
        "displayName": "name",
        "environment": Object {
          "category": "PRODUCTION",
          "displayName": "prod",
          "environmentId": "id",
        },
        "lastUpdated": "mock",
        "pipeline": Object {
          "displayName": "pipeline",
          "pipelineId": "1",
          "url": "url",
        },
        "sequenceNumber": 1,
        "state": "SUCCESSFUL",
        "updateSequenceNumber": "1",
        "url": "url",
      },
    ],
  },
}
`;
